as_strided symbolic support

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85260
* __->__ #85264

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D39662820](https://our.internmc.facebook.com/intern/diff/D39662820)

ezyang committed:
as_strided symbolic support

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

wconstab commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -12752,7 +12753,7 @@
+- func: as_strided_copy(Tensor self, SymInt[] size, SymInt[] stride, SymInt? storage_offset=None) -> Tensor
   variants: function
   dispatch:
     CompositeExplicitAutogradNonFunctional: as_strided_copy
```
ok i'm still a little confused about our naming conventions.  for implicit or other kernels i believe we landed on not changing the kernel name as codegen would do it for you.  but why not add `_symint` suffix here?

ezyang replied:
Simple: I didn't actually add symint support for as_strided_copy haha

ezyang committed:
Update on "as_strided symbolic support"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

wconstab reviewed (approved):
lgtm

ezyang committed:
Update on "as_strided symbolic support"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "as_strided symbolic support"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "as_strided symbolic support"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "as_strided symbolic support"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "as_strided symbolic support"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "as_strided symbolic support"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "as_strided symbolic support"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

Krovatkin commented on torch/csrc/lazy/core/ir_builder.h:
```diff
@@ -269,8 +269,10 @@ static inline NodePtr MakeSizeDiv(const Value& a, const Value& b) {
   return Value(
-      dynamic_cast<torch::lazy::SymIntNodeImpl*>(a.toSymIntNodeImpl().get())
-          ->node_,
+      a.is_symbolic() ? dynamic_cast<torch::lazy::SymIntNodeImpl*>(
```
I'd suggest to replace this `expect_int` I think we all agreed (cc: @wconstab ) we aren't going to create real symints in LTC. XLA won't be using codegen either

wconstab replied:
I guess XLA will be overriding this function? then I am ok with `expect_int`.  the way it is seems ok too, i guess it serves as a guide for how to handle symints in XLA but also confuses people since the rest of TS backend doesn't have DS functionality

ezyang replied:
I'm going to try not to wobble this code for now, if we want to rip out more of the embryonic LTC dynamic shapes stuff, let's do it in a separate pr

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39662820).

ezyang committed:
Update on "as_strided symbolic support"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D39662820](https://our.internmc.facebook.com/intern/diff/D39662820)

[ghstack-poisoned]

ezyang committed:
Update on "as_strided symbolic support"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

Differential Revision: [D39662820](https://our.internmc.facebook.com/intern/diff/D39662820)

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -f "windows queue is bad but we know it passes from prior runs"
